src/Tasks/LegalProposition/PrintIllegal.hs

```diff
@@ -10,53 +11,57 @@ import Trees.Helpers (collectLeaves, treeNodes)
 import Trees.Print (normalShow)
 import Trees.Types (BinOp(..), SynTree(..), allBinaryOperators, showOperator, showOperatorNot)
 import Data.List.Extra (replace)
+import Tasks.LegalProposition.Config (PropErrorReason (..))
+import Data.Bifunctor (Bifunctor(second, bimap))
+import Data.Maybe (fromJust, catMaybes)
 
 
 
 
-illegalDisplay :: SynTree BinOp Char -> Gen String
-illegalDisplay (Leaf _) = elements (showOperatorNot : map showOperator allBinaryOperators)
+illegalDisplay :: SynTree BinOp Char -> Gen (String, PropErrorReason)
+illegalDisplay (Leaf _) = (,IllegalOperand) <$> elements (showOperatorNot : map showOperator allBinaryOperators)
 illegalDisplay synTree =
     let usedLiterals = collectLeaves synTree
-    in replace' <$> ifUseIllegal True False synTree usedLiterals
+    in bimap replace' fromJust <$> ifUseIllegal True False synTree usedLiterals
       where replace' = replace "_" "" . replace "_ " "" . replace " _" ""
 
-ifUseIllegal :: Bool -> Bool -> SynTree BinOp Char -> String -> Gen String
+ifUseIllegal :: Bool -> Bool -> SynTree BinOp Char -> String -> Gen (String, Maybe PropErrorReason)
 ifUseIllegal useBug notFirstLayer synTree usedLiterals =
     let
       nodeNum = treeNodes synTree
     in
       if not useBug
-        then return (normalShow synTree)
-        else frequency
+        then return (normalShow synTree, Nothing)
+        else second Just <$> frequency
           [ (1, implementIllegal notFirstLayer synTree usedLiterals)
           , (fromIntegral nodeNum - 1, subTreeIllegal notFirstLayer synTree usedLiterals)
           ]
 
 
 
-subTreeIllegal ::Bool -> SynTree BinOp Char -> String -> Gen String
+subTreeIllegal ::Bool -> SynTree BinOp Char -> String -> Gen (String, PropErrorReason)
 subTreeIllegal notFirstLayer (Binary operator a b) usedLiterals =
     allocateBugToSubtree notFirstLayer a b usedLiterals operator
 subTreeIllegal _ (Not a) usedLiterals = do
     left <- ifUseIllegal True True a usedLiterals
-    return (showOperatorNot ++ left)
+    return $ bimap (showOperatorNot ++) fromJust left
 subTreeIllegal _ (Leaf _) _ = error "This will not happen but must be write"
 
 
 
-allocateBugToSubtree :: Bool -> SynTree BinOp Char -> SynTree BinOp Char -> String -> BinOp -> Gen String
+allocateBugToSubtree :: Bool -> SynTree BinOp Char -> SynTree BinOp Char -> String -> BinOp -> Gen (String, PropErrorReason)
 allocateBugToSubtree notFirstLayer a b usedLiterals usedOperator = do
     ifUseBug <- elements [True, False]
-    left <- ifUseIllegal ifUseBug True a usedLiterals
-    right <- ifUseIllegal (not ifUseBug) True b usedLiterals
+    (left, er1) <- ifUseIllegal ifUseBug True a usedLiterals
+    (right, er2) <- ifUseIllegal (not ifUseBug) True b usedLiterals
+    let errorReason = head $ catMaybes [er1,er2]
```

Comment:
GHC complains about the partial function `head` here. I find the construct `head . catMaybes` itself dubious. Better to at least directly use the `Alternative` instance for `Maybe`.